### ISSUE TITLE
Slot support?

### ISSUE THREAD

**Issue by mateomurphy** [camertron/rux#9]
I was just thinking it would be nice if slots were supported, so that we could do something like:

```
<BlogComponent>
  <slot name="header">
    {link_to "My blog", root_path}
  </slot>

  {BlogPost.all.each do |blog_post|
    <slot name="post">
      {link_to blog_post.name, blog_post.url}
    </slot>
  end}
</BlogComponent>
```

I'd be willing to help work on this if there's any interest?

**Comment by camertron** [comment]
Hey @mateomurphy, thanks for bringing this up 😄 I'm definitely in favor of adding slot support to rux.

As things stand today, slots already sort of work because they're supported by ViewComponent. For instance, this is currently possible:

```ruby
class ChildComponent < ViewComponent::Base
  renders_one :header, Header
  renders_many :items, Item

  def call
    <div>
      {header}
      {items.each do |item|
        <div>{item}</div>
      end}
    </div>
  end
end

class ParentComponent < ViewComponent::Base
  def call
    <ChildComponent as={child}>
      {child.with_header { "Header" }}
      {child.with_item { "Item 1" }}
      {child.with_item { "Item 2" }}
    </ChildComponent>
  end
end
```

So as you can see, we're almost there!

In playing with this just now, I've realized a couple of things:

1. Items from a `renders_many` slot can only be emitted if they're wrapped in an HTML element like I've done with `<div>` above. This is because, unlike ERB, Ruby code blocks have to be valid Ruby and there's no way to emit a string without first dropping back into HTML. Your idea of introducing a special `<slot>` element would help with this, although it would be used to render slots, not fill them in.
2. The `as={...}` feature is completely undocumented, woops 😅

Given all that, I think this is the API I would like to see:

```ruby
class ChildComponent < ViewComponent::Base
  renders_one :header, Header
  renders_many :items, Item

  def call
    <div>
      <Slot content={header} />
      {items.each do |item|
        <Slot content={item} />
      end}
    </div>
  end
end
```

For consistency, the slot itself should probably be passed in as some attribute (I chose `content`) rather than passing the name of the slot, since I can't think of a way to handle both regular (`renders_one`) and collection (`renders_many`) slots without making a mess of the API.

I _think_ this wouldn't be too difficult to implement, since `<Slot>` can just be another component that renders (i.e. calls `to_s` on) the value of the `contents` attribute, which will be an instance of `ViewComponent::Slot`.

Thoughts?

**Comment by mateomurphy** [comment]
Sorry just noticed your response.

> The as={...} feature is completely undocumented, woops 

Indeed I would not have asked this question if I knew about `as` :)

In any case, my example wasn't clear enough, as I'm suggesting a `Slot` component that can be used in the parent, the way they work in web components:

```ruby
class ParentComponent < ViewComponent::Base
  def call
    <ChildComponent>
      <Slot name="header">Header</Slot>
      <Slot name="item">Item 1</Slot>
      <Slot name="item">Item 2</Slot>
    </ChildComponent>
  end
end

# or

class ParentComponent < ViewComponent::Base
  def call
    <ChildComponent>
      <header slot="header">Header</header>
      <div slot="item">Item 1</slot>
      <div slot="item">Item 2</slot>
    </ChildComponent>
  end
end
```

My idea was that this would then be transpiled to the block syntax that `ViewComponent` supports.

In the meantime, I've implemented this in `ViewComponent` directly by adding react-like context mechanism, but it's surely not as efficient as supporting this directly in rux.

Also, regarding this:

> This is because, unlike ERB, Ruby code blocks have to be valid Ruby and there's no way to emit a string without first dropping back into HTML.

This could be mitigated by supporting JSX like fragments, like:

```ruby
class ChildComponent < ViewComponent::Base
  renders_one :header, Header
  renders_many :items, Item

  def call
    <div>
      {header}
      {items.each do |item|
        <>{item}</>
      end}
    </div>
  end
end
```

**Comment by camertron** [comment]
> Indeed I would not have asked this question if I knew about `as` :)

Yeah, that's definitely my bad. When slots are fully supported, I'd be happy to add a section to the README and mention the `as` argument.

> In any case, my example wasn't clear enough, as I'm suggesting a `Slot` component that can be used in the parent, the way they work in web components

Ah ok, I get it now. I hadn't considered the parallel with web components, that makes `Slot` a more attractive option.

> My idea was that this would then be transpiled to the block syntax that ViewComponent supports.

Right yeah, that makes sense 👍 

> This could be mitigated by supporting JSX like fragments

Great idea! We'd have to modify the rux parser to support fragment syntax, but it's definitely possible.

> I'd be willing to help work on this if there's any interest?

I'm on board with the features as you've described them here, namely:

1. Slots filled in with this syntax: `<Slot name="header">Header</Slot>`
2. Fragment support

It occurs to me that we may be able to remove `as` as well, so that's nice.

**Comment by camertron** [comment]
Hey @mateomurphy, I decided to take this on and just merged my changes and released v1.2.0. Check out the [README](https://github.com/camertron/rux#slots) for an example of the API I ultimately chose. It's a bit different than what we talked about here, but I'm pretty happy with how it turned out 😄
